<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Année 2020-2021 - 2e année département Sciences du Numérique </h1>
<h1> Noms: CHELLAF EL HAMMOUD </h1>
<h1> Prénoms: Chaimae </h1>    
</center>

# Algorithme de Newton
## Implémentation 
 
1. Coder l’algorithme de Newton local tel que décrit dans la section *Algorithme de Newton* (fichier `Algorithme_De_Newton.jl`)

2. Tester l’algorithme sur les fonctions $f_{1}$ , $f_{2}$ avec les points initiaux $x_{011}$ , $x_{012}$ (pour $f_{1}$ ) et $x_{021}$ , $x_{022}$ , $x_{023}$ (pour $f_{2}$ ) donnés en Annexe A.

In [12]:
using Pkg
Pkg.status()

Status `/applications/julia/.julia/environments/v1.4/Project.toml`
  [13f3f980] CairoMakie v0.6.4
  [5ae59095] Colors v0.12.8
  [a81c6b42] Compose v0.9.2
  [0c46a032] DifferentialEquations v6.18.0
  [e30172f5] Documenter v0.27.10
  [587475ba] Flux v0.11.1
  [f6369f11] ForwardDiff v0.10.23
  [28b8d3ca] GR v0.52.0
  [c91e804a] Gadfly v1.3.4
  [a2cc645c] GraphPlot v0.5.0
  [7073ff75] IJulia v1.23.2
  [b6b21f68] Ipopt v0.8.0
  [4138dd39] JLD v0.12.3
  [033835bb] JLD2 v0.4.3
  [4076af6c] JuMP v0.22.1
  [093fc24a] LightGraphs v1.3.5
  [2774e3e8] NLsolve v4.5.1
  [91a5bcdd] Plots v1.6.12
  [2913bbd2] StatsBase v0.33.13
  [e88e6eb3] Zygote v0.5.17


In [13]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub
include("Algorithme_De_Newton.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,nom_fct,point_init,xmin,fxmin,flag,sol_exacte,nbiters)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",nom_fct, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xsol = ",xmin)
	println("  * f(xsol) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * flag = ",flag)
	println("  * sol_exacte : ", sol_exacte)
end


# Fonction f0
# -----------
f0(x) =  sin(x)
# la gradient de la fonction f0
grad_f0(x) = cos(x)
# la hessienne de la fonction f0
hess_f0(x) = -sin(x)
sol_exacte = -pi/2
options = []

x0 = sol_exacte
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = -pi/2+0.5
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = pi/2
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)

# Annexe A

# Les points initiaux
# pour les problèmes sans contraintes
struct Pts_sans_contraintes
    x011
    x012
    x021
    x022
    x023
end
x011 = [1; 0; 0]
x012 = [10; 3; -2.2]
x021 = [-1.2; 1]
x022 = [10; 0]
x023 = [0; 1/200 + 1/10^12]
# les points initiaux utilisés dans les problèmes sans contraintes
pts1 = Pts_sans_contraintes(x011,x012,x021,x022,x023)

# La première fonction de test
fct1(x) =  2*(x[1]+x[2]+x[3]-3)^2 + (x[1]-x[2])^2 + (x[2]-x[3])^2
#Le gradient de la fonction fct1
function grad_fct1(x)
    y1 = 4*(x[1]+x[2]+x[3]-3) + 2*(x[1]-x[2])
    y2 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[1]-x[2]) +2*(x[2]-x[3])
    y3 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[2]-x[3])
    return [y1;y2;y3]
end
# la hessienne de la fonction fct1
hess_fct1(x) = [6 2 4;2 8 2;4 2 6]

# La deuxième fonction de test
fct2(x)=100*(x[2]-x[1]^2)^2+(1-x[1])^2
# la gradient de la fonction fct2
grad_fct2(x)=[-400*x[1]*(x[2]-x[1]^2)-2*(1-x[1]) ; 200*(x[2]-x[1]^2)]
#la hessienne de la fonction fct2
hess_fct2(x)=[-400*(x[2]-3*x[1]^2)+2  -400*x[1];-400*x[1]  200]

sol_exacte_fct1 = [1;1;1]
sol_exacte_fct2 = [1;1]
options = []

xmin,f_min,flag,nb_iters = Algorithme_De_Newton(fct1,grad_fct1,hess_fct1,pts1.x011,options)
my_afficher_resultats("Newton","f1",pts1.x011,xmin,f_min,flag,sol_exacte_fct1,nb_iters)
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(fct1,grad_fct1,hess_fct1,pts1.x012,options)
my_afficher_resultats("Newton","f1",pts1.x012,xmin,f_min,flag,sol_exacte_fct1,nb_iters)

xmin,f_min,flag,nb_iters = Algorithme_De_Newton(fct2,grad_fct2,hess_fct2,pts1.x021,options)
my_afficher_resultats("Newton","f2",pts1.x021,xmin,f_min,flag,sol_exacte_fct2,nb_iters)
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(fct2,grad_fct2,hess_fct2,pts1.x022,options)
my_afficher_resultats("Newton","f2",pts1.x022,xmin,f_min,flag,sol_exacte_fct2,nb_iters)
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(fct2,grad_fct2,hess_fct2,pts1.x023,options)
my_afficher_resultats("Newton","f2",pts1.x023,xmin,f_min,flag,sol_exacte_fct2,nb_iters)



-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.5707963267948966:
  * xsol = -1.5707963267948966
  * f(xsol) = -1.0
  * nb_iters = 1
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.0707963267948966:
  * xsol = -1.5707963267949088
  * f(xsol) = -1.0
  * nb_iters = 3
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial 1.5707963267948966:
  * xsol = 1.5707963267948966
  * f(xsol) = 1.0
  * nb_iters = 1
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f1 au point initial [1, 0, 0]:
  * xsol = [1.0, 1.0, 0.9999999999999999]
  * f(xsol) = 1.232595164407

SingularException: SingularException(2)

## Interprétation 

Justifier

1. les résultats obtenus pour l'exemple $f_0$ ci-dessus;

2. que l’algorithme implémenté converge en une itération pour $f_{1}$;

3. que l’algorithme puisse ne pas converger pour $f_{2}$ avec certains points initiaux.

## Vos réponses?

1- Pour les points initiaux -1.5707963267948966 et -1.070796326794896, l'algorithme converge vers la solution qui est bien correcte puisque cette valeur correspond au minimum de la fonction sinus. Cependant, le troisième test montre qu'il n y a pas convergence vers le minimum mais plutôt vers le maximum de sinus, vu que, en partant du point initial 1.5707963267948966, l'approximation à l'ordre deux de Taylor sur f0 se dirige vers le sens opposé à un minimum, d'où le résultat. 

2- f1 est égale à son développement de Taylor à l'ordre deux, et donc après une seule itération, elle atteint son minimum (vu que le gradient en ce point après la première itération est nul).

3- Comme illustré avec le point de départ x023 = [0; 1/200 + 1/10^12] sur le test de la fonction f2, l'algorithme de Newton n'a pas convergé. Donc, cet algorithme peut diverger pour f2 avec certains points initiaux. On peut justifier ceci par le fait que notre problème peut ne pas avoir de solutions si la Hessienne de f dans le point xk n'est pas inversible.

# Régions de confiance avec pas de cauchy 

## Implémentation 

1. Coder l'algorithme du pas de Cauchy d’un sous-problème de
régions de confiance (fichier `Pas_De_Cauchy.jl`). Tester sur les quadratiques proposées en Annexe B.

2. Coder l'algorithme de régions de confiance (fichier `Regions_De_Confiance.jl`). Tester sur les problèmes de l’Annexe A.

In [1]:
using LinearAlgebra
using Markdown  

include("Pas_De_Cauchy.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,g, H, delta, s, e)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à g = ",g, " et H = ", H, " avec delta = " , delta, ":\n",bold=true,color=:blue)
	println("  * s = ",s)
	println("  * e = ",e)
end


# Vos tests

g1 = [0; 0]
H1 = [7 0 ; 0 2]
delta1 = 1
s1, e1 = Pas_De_Cauchy(g1,H1,delta1)

g2 = [6; 2]
H2 = [7 0 ; 0 2]
delta2 = 1
s2, e2 = Pas_De_Cauchy(g2,H2,delta2)

g3 = [-2; 1]
H3 = [-2 0 ; 0 10]
delta3 = 1
s3, e3 = Pas_De_Cauchy(g3,H3,delta3)

my_afficher_resultats("Pas de Cauchy",g1, H1, delta1, s1, e1)
my_afficher_resultats("Pas de Cauchy",g2, H2, delta2, s2, e2)
my_afficher_resultats("Pas de Cauchy",g3, H3, delta3, s3, e3)







-------------------------------------------------------------------------
Résultats de : Pas de Cauchy appliqué à g = [0, 0] et H = [7 0; 0 2] avec delta = 1:
  * s = [0.0, 0.0]
  * e = 0
-------------------------------------------------------------------------
Résultats de : Pas de Cauchy appliqué à g = [6, 2] et H = [7 0; 0 2] avec delta = 1:
  * s = [-0.9230769230769231, -0.3076923076923077]
  * e = 1
-------------------------------------------------------------------------
Résultats de : Pas de Cauchy appliqué à g = [-2, 1] et H = [-2 0; 0 10] avec delta = 1:
  * s = [0.8944271909999159, -0.4472135954999579]
  * e = -1


In [11]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub
include("Pas_De_Cauchy.jl")
include("Regions_De_Confiance.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,fct,point_init, xmin, fxmin, nbiters, sol_exacte)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",fct, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xmin = ",xmin)
	println("  * f(xmin) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * sol_exacte : ", sol_exacte)
end

#Tests

# La tolérance utilisée dans les tests
	tol_erreur = 1e-2
	# initialisation des variables de l'algorithme
	gamma1 = 0.5
	gamma2 = 2.00
	eta1 = 0.25
	eta2 = 0.75
	deltaMax = 10
	Tol_abs = sqrt(eps())
	Tol_rel = 1e-8
	maxits = 1000     # on peut mettre maxits = 5000 pour voir que tous les résultats convergent
	delta0_1 = 2
	delta0_2 = 2
	options1 =[deltaMax,gamma1,gamma2,eta1,eta2,delta0_1,maxits,Tol_abs,Tol_rel]
	options2 =[deltaMax,gamma1,gamma2,eta1,eta2,delta0_2,maxits,Tol_abs,Tol_rel]

    # Les points initiaux
    # pour les problèmes sans contraintes
    struct Pts_sans_contraintes
        x011
        x012
        x021
        x022
        x023
    end
    x011 = [1; 0; 0]
    x012 = [10; 3; -2.2]
    x021 = [-1.2; 1]
    x022 = [10; 0]
    x023 = [0; 1/200 + 1/10^12]
    # les points initiaux utilisés dans les problèmes sans contraintes
    pts1 = Pts_sans_contraintes(x011,x012,x021,x022,x023)

    # La première fonction de test
    fct1(x) =  2*(x[1]+x[2]+x[3]-3)^2 + (x[1]-x[2])^2 + (x[2]-x[3])^2
    #Le gradient de la fonction fct1
    function grad_fct1(x)
        y1 = 4*(x[1]+x[2]+x[3]-3) + 2*(x[1]-x[2])
        y2 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[1]-x[2]) +2*(x[2]-x[3])
        y3 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[2]-x[3])
        return [y1;y2;y3]
    end
    # la hessienne de la fonction fct1
    hess_fct1(x) = [6 2 4;2 8 2;4 2 6]

    # La deuxième fonction de test
    fct2(x)=100*(x[2]-x[1]^2)^2+(1-x[1])^2
    # la gradient de la fonction fct2
    grad_fct2(x)=[-400*x[1]*(x[2]-x[1]^2)-2*(1-x[1]) ; 200*(x[2]-x[1]^2)]
    #la hessienne de la fonction fct2
    hess_fct2(x)=[-400*(x[2]-3*x[1]^2)+2  -400*x[1];-400*x[1]  200]

    sol_exacte_fct1 = [1;1;1]
    sol_exacte_fct2 = [1;1]
    
x_min11, fmin11, flag11, nb_iters11 = Regions_De_Confiance("cauchy",fct1,grad_fct1,hess_fct1,pts1.x011,options1)
my_afficher_resultats("Regions de Confiance CAUCHY",fct1,pts1.x011, x_min11, fmin11, nb_iters11, sol_exacte_fct1)
x_min12, fmin12, flag12, nb_iters12 = Regions_De_Confiance("cauchy",fct1,grad_fct1,hess_fct1,pts1.x012,options1)
my_afficher_resultats("Regions de Confiance CAUCHY",fct1,pts1.x012, x_min12, fmin12, nb_iters12, sol_exacte_fct1)



x_min21, fmin21, flag21, nb_iters21,  = Regions_De_Confiance("cauchy",fct2,grad_fct2,hess_fct2,pts1.x021,options2)
my_afficher_resultats("Regions de Confiance CAUCHY",fct2,pts1.x021, x_min21, fmin21, nb_iters21, sol_exacte_fct2)
x_min22, fmin22, flag22, nb_iters22 = Regions_De_Confiance("cauchy",fct2,grad_fct2,hess_fct2,pts1.x022,options2)
my_afficher_resultats("Regions de Confiance CAUCHY",fct2,pts1.x022, x_min22, fmin22, nb_iters22, sol_exacte_fct2)
x_min23, fmin23, flag23, nb_iters23 = Regions_De_Confiance("cauchy",fct2,grad_fct2,hess_fct2,pts1.x023,options2)
my_afficher_resultats("Regions de Confiance CAUCHY",fct2,pts1.x023, x_min23, fmin23, nb_iters23, sol_exacte_fct2)


-------------------------------------------------------------------------
Résultats de : Regions de Confiance CAUCHY appliqué à fct1 au point initial [1, 0, 0]:
  * xmin = [1.0000558873349883, 0.999992420017735, 0.9999289527004819]
  * f(xmin) = 9.090411079109608e-9
  * nb_iters = 26
  * sol_exacte : [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : Regions de Confiance CAUCHY appliqué à fct1 au point initial [10.0, 3.0, -2.2]:
  * xmin = [1.0000497954627436, 0.9999961002424802, 0.9999424049876048]
  * f(xmin) = 6.040104651861e-9
  * nb_iters = 28
  * sol_exacte : [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : Regions de Confiance CAUCHY appliqué à fct2 au point initial [-1.2, 1.0]:
  * xmin = [0.9124243892242884, 0.8318955539092286]
  * f(xmin) = 0.007708264643929195
  * nb_iters = 1001
  * sol_exacte : [1, 1]
-------------------------------------------------------------------------
Ré

## Interprétation 

1. Quelle relation lie la fonction test $f_1$ et son modèle de Taylor à l’ordre 2 ? Comparer alors les performances de Newton et RC-Pas de Cauchy sur cette fonction.

2. Le rayon initial de la région de confiance est un paramètre important dans l’analyse
de la performance de l’algorithme. Sur quel(s) autre(s) paramètre(s) peut-on jouer
pour essayer d’améliorer cette performance ? Étudier l’influence d’au moins deux de
ces paramètres.

## Vos réponses?

1- f1 est égale à son modèle de Taylor à l'ordre 2.
Pour l'algorithme de Newton, la fonction f1 converge en une seule itérationn alors que pour l'algorithme de régions de confiance avec le pas de cauchy, le code fait plus de 10 itérations pour trouver la solution. Donc, l'algorithme de Newton est, dans ce cas, moins couteux et plus efficace vu qu'il donne les valeurs exactes en un nombre d'itérations plus petit.

2- Pour analyser la performance de l'algorithme, on peut jouer sur plusieurs paramètres comme : 

-> η1, η2 : les critères d'agrandissement et réduction de la région de confiance. Si on diminue lintervalle [η1, η2], on diminue la chance de trouver ρk, et donc on fera plus d'itérations pour résoudre le problème.

-> γ1, γ2 : les facteurs d'agrandissement et de réduction de la région de confiance. Plus ces facteurs sont petits, plus la nouvelle région de confiance Δk+1 est petite.

-> Δmax : le rayon de confiance maximale.


# Régions de confiance avec gradient conjugué tronqué

## Implémentation 

1. Implémenter l’algorithme du Gradient Conjugué Tronqué, en se basant sur le cours (fichier `Gradient_Conjugue_Tronque.jl`).
On validera les résultats sur les fonctions de l’Annexe C.

2. Intégrer finalement l’algorithme du Gradient Conjugué Tronqué dans le code de
régions de confiance, et appliquer ce code pour résoudre les exemples proposés en
Annexe A.

In [12]:
using LinearAlgebra
using Markdown  

include("Pas_De_Cauchy.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,g, H, s)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à g = ",g, " et H = ", H, ":\n",bold=true,color=:blue)
	println("  * s = ",s)
end


# Vos tests

g4 = [0; 0]
H4 = [-2 0 ; 0 10]
s4 = Gradient_Conjugue_Tronque(g4,H4,[])

g5 = [2; 3]
H5 = [4 6 ; 6 5]
s5 = Gradient_Conjugue_Tronque(g5,H5,[])

g6 = [2; 0]
H6 = [4 0 ; 0 -15]
s6 = Gradient_Conjugue_Tronque(g6, H6,[])

my_afficher_resultats("Gradient Conjugue Tronque",g4, H4, s4)
my_afficher_resultats("Gradient Conjugue Tronque",g5, H5, s5)
my_afficher_resultats("Gradient Conjugue Tronque",g6, H6, s6)




-------------------------------------------------------------------------
Résultats de : Gradient Conjugue Tronque appliqué à g = [0, 0] et H = [-2 0; 0 10]:
  * s = [0.0, 0.0]
-------------------------------------------------------------------------
Résultats de : Gradient Conjugue Tronque appliqué à g = [2, 3] et H = [4 6; 6 5]:
  * s = [1.1782448197996298, -1.6160876042514951]
-------------------------------------------------------------------------
Résultats de : Gradient Conjugue Tronque appliqué à g = [2, 0] et H = [4 0; 0 -15]:
  * s = [-0.5, 0.0]


In [13]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub
include("Gradient_Conjugue_Tronque.jl")
include("Regions_De_Confiance.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,fct,point_init, xmin, fxmin, nbiters, sol_exacte)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",fct, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xmin = ",xmin)
	println("  * f(xmin) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * sol_exacte : ", sol_exacte)
end

#Tests

# La tolérance utilisée dans les tests
	tol_erreur = 1e-2
	# initialisation des variables de l'algorithme
	gamma1 = 0.5
	gamma2 = 2.00
	eta1 = 0.25
	eta2 = 0.75
	deltaMax = 10
	Tol_abs = sqrt(eps())
	Tol_rel = 1e-8
	maxits = 5000
	delta0_1 = 2
	delta0_2 = 2
	options1 =[deltaMax,gamma1,gamma2,eta1,eta2,delta0_1,maxits,Tol_abs,Tol_rel]
	options2 =[deltaMax,gamma1,gamma2,eta1,eta2,delta0_2,maxits,Tol_abs,Tol_rel]

    # Les points initiaux
    # pour les problèmes sans contraintes
    struct Pts_sans_contraintes
        x011
        x012
        x021
        x022
        x023
    end
    x011 = [1; 0; 0]
    x012 = [10; 3; -2.2]
    x021 = [-1.2; 1]
    x022 = [10; 0]
    x023 = [0; 1/200 + 1/10^12]
    # les points initiaux utilisés dans les problèmes sans contraintes
    pts1 = Pts_sans_contraintes(x011,x012,x021,x022,x023)

    # La première fonction de test
    fct1(x) =  2*(x[1]+x[2]+x[3]-3)^2 + (x[1]-x[2])^2 + (x[2]-x[3])^2
    #Le gradient de la fonction fct1
    function grad_fct1(x)
        y1 = 4*(x[1]+x[2]+x[3]-3) + 2*(x[1]-x[2])
        y2 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[1]-x[2]) +2*(x[2]-x[3])
        y3 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[2]-x[3])
        return [y1;y2;y3]
    end
    # la hessienne de la fonction fct1
    hess_fct1(x) = [6 2 4;2 8 2;4 2 6]

    # La deuxième fonction de test
    fct2(x)=100*(x[2]-x[1]^2)^2+(1-x[1])^2
    # la gradient de la fonction fct2
    grad_fct2(x)=[-400*x[1]*(x[2]-x[1]^2)-2*(1-x[1]) ; 200*(x[2]-x[1]^2)]
    #la hessienne de la fonction fct2
    hess_fct2(x)=[-400*(x[2]-3*x[1]^2)+2  -400*x[1];-400*x[1]  200]

    sol_exacte_fct1 = [1;1;1]
    sol_exacte_fct2 = [1;1]
    
x_min11, fmin11, flag11, nb_iters11 = Regions_De_Confiance("gct",fct1,grad_fct1,hess_fct1,pts1.x011,options1)
my_afficher_resultats("Regions de Confiance GCT",fct1,pts1.x011, x_min11, fmin11, nb_iters11, sol_exacte_fct1)
x_min12, fmin12, flag12, nb_iters12 = Regions_De_Confiance("gct",fct1,grad_fct1,hess_fct1,pts1.x012,options1)
my_afficher_resultats("Regions de Confiance GCT",fct1,pts1.x012, x_min12, fmin12, nb_iters12, sol_exacte_fct1)



x_min21, fmin21, flag21, nb_iters21,  = Regions_De_Confiance("gct",fct2,grad_fct2,hess_fct2,pts1.x021,options2)
my_afficher_resultats("Regions de Confiance GCT",fct2,pts1.x021, x_min21, fmin21, nb_iters21, sol_exacte_fct2)
x_min22, fmin22, flag22, nb_iters22 = Regions_De_Confiance("gct",fct2,grad_fct2,hess_fct2,pts1.x022,options2)
my_afficher_resultats("Regions de Confiance GCT",fct2,pts1.x022, x_min22, fmin22, nb_iters22, sol_exacte_fct2)
x_min23, fmin23, flag23, nb_iters23 = Regions_De_Confiance("gct",fct2,grad_fct2,hess_fct2,pts1.x023,options2)
my_afficher_resultats("Regions de Confiance GCT",fct2,pts1.x023, x_min23, fmin23, nb_iters23, sol_exacte_fct2)


-------------------------------------------------------------------------
Résultats de : Regions de Confiance GCT appliqué à fct1 au point initial [1, 0, 0]:
  * xmin = [1.0000000000000007, 1.0, 1.0]
  * f(xmin) = 2.0214560696288428e-30
  * nb_iters = 1
  * sol_exacte : [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : Regions de Confiance GCT appliqué à fct1 au point initial [10.0, 3.0, -2.2]:
  * xmin = [0.9999999999999996, 1.0000000000000004, 1.0000000000000009]
  * f(xmin) = 2.5637979419682884e-30
  * nb_iters = 3
  * sol_exacte : [1, 1, 1]
-------------------------------------------------------------------------
Résultats de : Regions de Confiance GCT appliqué à fct2 au point initial [-1.2, 1.0]:
  * xmin = [0.999999674378009, 0.9999993478371609]
  * f(xmin) = 1.0611413034983129e-13
  * nb_iters = 31
  * sol_exacte : [1, 1]
-------------------------------------------------------------------------
Résultats de : Regions de Confiance 

## Interprétation  

1. Comparer la décroissance obtenue avec celle du pas de Cauchy, en retournant, dans
un premier temps le dernier itéré admissible à courbure positive (c’est à dire, que si
l’une ou l’autre des deux conditions (b) ou (d) sont rencontrées dans l’algorithme 3,
alors on ne calcule pas ``σ_{j}`` et on retourne le dernier itéré ``s_{j}`` directement).

2. Comparer la décroissance obtenue avec celle du pas de Cauchy, en imposant la sortie
dans l’algorithme 3 au bout d’une itération seulement. Que remarquez vous ?

3. Comparer la décroissance obtenue avec celle du pas de Cauchy dans le cas général.

4. Quels sont les avantages et inconvénients des deux approches ?

## Vos réponses?

2- Après avoir comparé la décroissance obtenue dans cet algorithme avec celle du pas de cauchy, en imposant la sortie au bout d'une itération, le résultat était très proche. Il n y a pas donc trop de différence en travaillant en une seule itération.

3- Dans le cas général, la décroissance obtenue par cet algorithme est meilleure que celle obtenue par le pas de Cauchy. Cependant, il y a parfois des valeurs négatives de la décroissance en appliquant l'algorithme du gradient conjugué tronqué.

4-En se basant sur la décroissance ainsi que le nombre d'itérations obtenus à l'issue des exemples ci-dessus, l'algorithme du Pas de Cauchy est plus lent que l'algorithme du gradient conjugué, mais il ne donne que des valeurs positives de la décroissance. L'algorithme du gradient conjugué est donc plus rapide, or il peut donner des valeurs négatives, et donc il risque d'augmenter la valeur de notre fonction au lieu de la minimiser.



# Lagrangien augmenté

## Implémentation
1.Choisir des critères d’arrêt pour la convergence de l'algorithme.

2.Implémenter l'algorithme du lagrangien augmenté, en utilisant les différentes méthodes
qui ont été vues en première partie pour la résolution de la suite de problémes sans
contraintes (fichier `Lagrangien_Augmente.jl`)
 
3.Tester les différentes variantes sur les problèmes en Annexe D.
 

In [14]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub
include("Algorithme_De_Newton.jl")
include("Pas_De_Cauchy.jl")
include("Gradient_Conjugue_Tronque.jl")
include("Regions_De_Confiance.jl")
include("Lagrangien_Augmente.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,fct, contrainte, point_init, xmin, fxmin, nbiters, sol_exacte)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",fct, " avec la contrainte ", contrainte, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xmin = ",xmin)
	println("  * f(xmin) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * sol_exacte : ", sol_exacte)
end


#Tests
# initialisation des paramètres
	lambda0 = 2
	mu0 = 10
	tho = 2
	epsilon = 1e-8
	tol = 1e-5
	max_iters = 1000
	options = [epsilon, tol, max_iters, lambda0, mu0, tho]
	# La tolérance utilisée dans les tests
	tol_erreur = 1e-4

    # pour les problèmes avec contraintes
    struct Pts_avec_contraintes
        x01
        x02
        x03
        x04
    end
    x01 = [0; 1; 1]
    x02 = [0.5; 1.25; 1]
    x03 = [1; 0]
    x04 = [sqrt(3)/2 ;sqrt(3)/2]

    pts2 = Pts_avec_contraintes(x01,x02,x03,x04)

    #Les solutions exactes
    sol_fct1_augm = [0.5 ; 1.25 ; 0.5]
    sol_fct2_augm = [0.9072339605110892; 0.82275545631455]

    # Les fonctions 

    fct1(x) =  2*(x[1]+x[2]+x[3]-3)^2 + (x[1]-x[2])^2 + (x[2]-x[3])^2
    # la gradient de la fonction fct1
    function grad_fct1(x)
        y1 = 4*(x[1]+x[2]+x[3]-3) + 2*(x[1]-x[2])
        y2 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[1]-x[2]) +2*(x[2]-x[3])
        y3 = 4*(x[1]+x[2]+x[3]-3) - 2*(x[2]-x[3])
        return [y1;y2;y3]
    end
    # la hessienne de la fonction fct1
    hess_fct1(x) = [6 2 4;2 8 2;4 2 6]


    fct2(x)=100*(x[2]-x[1]^2)^2+(1-x[1])^2
    # la gradient de la fonction fct2
    grad_fct2(x)=[-400*x[1]*(x[2]-x[1]^2)-2*(1-x[1]) ; 200*(x[2]-x[1]^2)]
    #la hessienne de la fonction fct2
    hess_fct2(x)=[-400*(x[2]-3*x[1]^2)+2  -400*x[1];-400*x[1]  200]

    #Les contraintes
    contrainte1(x) =  x[1]+x[3]-1
    grad_contrainte1(x) = [1 ;0; 1]
    hess_contrainte1(x) = [0 0 0;0 0 0;0 0 0]

    contrainte2(x) =  (x[1]^2) + (x[2]^2) -1.5
    grad_contrainte2(x) = [2*x[1] ;2*x[2]]
    hess_contrainte2(x) = [2 0;0 2]
	
	# Les trois algorithmes d'optimisations sans contraintes utlisés
	algos = ["newton", "gct", "cauchy"]

    for algo in algos
        xmin,fxmin,flag,nbiters = Lagrangien_Augmente(algo,fct1,contrainte1,grad_fct1,hess_fct1,grad_contrainte1,
			hess_contrainte1,pts2.x01,options)
        my_afficher_resultats(algo,fct1, contrainte1, pts2.x01, xmin, fxmin, nbiters, sol_fct1_augm)
        xmin ,fxmin,flag,nbiters = Lagrangien_Augmente(algo,fct1,contrainte1,grad_fct1,hess_fct1,grad_contrainte1,
			hess_contrainte1,pts2.x02,options)
        my_afficher_resultats(algo,fct1, contrainte1, pts2.x02, xmin, fxmin, nbiters, sol_fct1_augm)
        xmin,fxmin,flag,nbiters = Lagrangien_Augmente(algo,fct2,contrainte2,grad_fct2,hess_fct2,grad_contrainte2,
			hess_contrainte2,pts2.x03,options)
        my_afficher_resultats(algo,fct2, contrainte2, pts2.x03, xmin, fxmin, nbiters, sol_fct2_augm)
        xmin ,fxmin,flag,nbiters = Lagrangien_Augmente(algo,fct2,contrainte2,grad_fct2,hess_fct2,grad_contrainte2,
			hess_contrainte2,pts2.x04,options)
        my_afficher_resultats(algo,fct2, contrainte2, pts2.x04, xmin, fxmin, nbiters, sol_fct2_augm)
    end


-------------------------------------------------------------------------
Résultats de : newton appliqué à fct1 avec la contrainte contrainte1 au point initial [0, 1, 1]:
  * xmin = [0.5000000061804462, 1.2499999969097768, 0.5000000061804464]
  * f(xmin) = 2.2499999443759835
  * nb_iters = 10
  * sol_exacte : [0.5, 1.25, 0.5]
-------------------------------------------------------------------------
Résultats de : newton appliqué à fct1 avec la contrainte contrainte1 au point initial [0.5, 1.25, 1.0]:
  * xmin = [0.5000000120317561, 1.2499999939841218, 0.5000000120317564]
  * f(xmin) = 2.2499998917141943
  * nb_iters = 9
  * sol_exacte : [0.5, 1.25, 0.5]
-------------------------------------------------------------------------
Résultats de : newton appliqué à fct2 avec la contrainte contrainte2 au point initial [1, 0]:
  * xmin = [0.9072339050834595, 0.8227553555256332]
  * f(xmin) = 0.008615660957332997
  * nb_iters = 4
  * sol_exacte : [0.9072339605110892, 0.82275545631455]
----------

## Interprétation
 1.Commenter les résultats obtenus, en étudiant notamment les valeurs de $\lambda_k$ et $\mu_k$.
 
 2.Étudier l'influence du paramètre $\tau$ dans la performance de l'algorithme.
 
 3.**Supplémentaire** : 
      Que proposez-vous comme méthode pour la résolution des problèmes avec
      des contraintes à la fois d'égalité et d'inégalité ? Implémenter (si le temps le permet)
      ce nouvel algorithme


## Vos réponses?

1- Les résultats obtenus sont similaires et ce sont les bons résultats attendus. 
Les valeurs de λk et μk sont aussi importants vu qu'ils sont bien liés au nombre d'itérations et notamment au critère de convergence global. Et d'après l'expression du Lagrangien, on peut voir que si on augmente la valeur de μk, on accélère l'algorithme (multiplication à un terme au carré), contrairement à λk qui n'a pas beaucoup d'influence sur l'algorithme malgré son importance. Dans ce sens, μk permet de résoudre des problèmes complexes qui prennent beaucoup de temps si on applique d'autres méthodes de résolution de problèmes. 

2- On peut voir d'après l'algorithme que τ est lié à μk sous une condition sur la contrainte, et donc plus on augmente τ, plus on augmente aussi μk, et comme expliqué dans la question précédente, ceci permet de trouver une meilleure convergence à l'algorithme. Cependant, on ne peut pas choisir une valeur de τ très petite (car sinon on perd en précision), et on ne peut pas choisir une valeur de τ qui soit très grande. On doit donc bien choisir une valeur adaptée à notre problème.

3- Afin de résoudre des problèmes avec contraintes à la fois d'égalité et d'inégalité, on peut exprimer les inégalités sous formes d'égalités (en ajoutant de nouveaux termes par exemple), et comme ça on se ramène à un algorithme similaire à le nôtre mais avec plus de contraintes et plus de termes.
